File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
                    
        if (head == null || head.next == null) {
            return true;
        }
        
        ListNode slow = head, fast = head, prev = null;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            ListNode nextTemp = slow.next;
            slow.next = prev;
            prev = slow;
            slow = nextTemp;
        }
        
        if (fast != null) {
            slow = slow.next; // Skip the middle element for odd-length lists
        }
        
        while (slow != null) {
            if (slow.val != prev.val) {
                return false;
            }
            slow = slow.next;
            prev = prev.next;
        }
        
        return true;
        
    }
}

File: 234-palindrome-linked-list/palindrome-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
int len(struct ListNode* head){
    int count =0;
    while(head!=NULL){
        count++;
        head = head->next;
    }
    return count;
}
bool isPalindrome(struct ListNode* head) {
    int n = len(head);
    int arr[n];
    struct ListNode* temp = head;
    for(int i=0;i<n;i++){
        arr[i]=temp->val;
        temp = temp->next;
    }

    for(int i=0;i<n;i++){
        if(arr[i]!=arr[n-i-1]){
            return false;

        }
    }
    return true;

    
    
}